Thumbnails taking much longer to load

Hi, I just rebuilt release-4.1.0 from source and I noticed this issue.

The thumbnails takes much longer to load in the media viewer, and I notice this error popping up in console logs which seems to be related:

```
card-0c55e731.js:75 Error: Failed to receive response from Home Assistant for request
    at Pp (card-0c55e731.js:141:10229)
    at async QM (card-0c55e731.js:320:22494)
    at async i (card-0c55e731.js:320:39222)
    at async Promise.all (/lovelace/index 0)
    at async pT.getEvents (card-0c55e731.js:320:39410)
    at async a (card-0c55e731.js:320:16313)
    at async Promise.all (/lovelace/index 0)
    at async o (card-0c55e731.js:320:16637)
    at async Promise.all (/lovelace/index 0)
    at async MM._handleQuery (card-0c55e731.js:320:16719) {request: {…}, response: {…}}
```

PS: I'm using Frigate 0.12.

--- a/src/camera-manager/engine-factory.ts
+++ b/src/camera-manager/engine-factory.ts
@@ -22,7 +22,7 @@ export class CameraManagerEngineFactory {
     this._cardWideConfig = cardWideConfig;
   }
 
-  public async createEngine(engine: Engine): Promise<CameraManagerEngine | null> {
+  public createEngine(engine: Engine): CameraManagerEngine | null {
     let cameraManagerEngine: CameraManagerEngine | null = null;
     switch (engine) {
       case Engine.Generic:

--- a/src/camera-manager/engine.ts
+++ b/src/camera-manager/engine.ts
@@ -22,6 +22,8 @@ import {
   CameraConfigs,
   Engine,
   CameraEndpoints,
+  MediaMetadataQuery,
+  MediaMetadataQueryResultsMap,
 } from './types';
 
 export const CAMERA_MANAGER_ENGINE_EVENT_LIMIT_DEFAULT = 10000;
@@ -106,7 +108,8 @@ export interface CameraManagerEngine {
   getMediaMetadata(
     hass: HomeAssistant,
     cameras: CameraConfigs,
-  ): Promise<MediaMetadata | null>;
+    query: MediaMetadataQuery,
+  ): Promise<MediaMetadataQueryResultsMap | null>;
 
   getCameraMetadata(
     hass: HomeAssistant,

--- a/src/camera-manager/frigate/engine-frigate.ts
+++ b/src/camera-manager/frigate/engine-frigate.ts
@@ -4,7 +4,7 @@ import endOfHour from 'date-fns/endOfHour';
 import startOfHour from 'date-fns/startOfHour';
 import { CameraConfig, CardWideConfig } from '../../types';
 import { ViewMedia } from '../../view/media';
-import { RequestCache, RecordingSegmentsCache } from '../cache';
+import { RecordingSegmentsCache, RequestCache } from '../cache';
 import {
   CameraManagerEngine,
   CAMERA_MANAGER_ENGINE_EVENT_LIMIT_DEFAULT,
@@ -19,9 +19,6 @@ import {
   EventQuery,
   EventQueryResults,
   EventQueryResultsMap,
-  FrigateEventQueryResults,
-  FrigateRecordingQueryResults,
-  FrigateRecordingSegmentsQueryResults,
   MediaMetadata,
   PartialEventQuery,
   PartialRecordingQuery,
@@ -40,8 +37,16 @@ import {
   CameraConfigs,
   CameraEndpoints,
   CameraEndpoint,
+  MediaMetadataQuery,
+  MediaMetadataQueryResults,
+  MediaMetadataQueryResultsMap,
 } from '../types';
-import { FrigateRecording } from './types';
+import {
+  FrigateEventQueryResults,
+  FrigateRecordingQueryResults,
+  FrigateRecordingSegmentsQueryResults,
+  FrigateRecording,
+} from './types';
 import {
   getEvents,
   getEventSummary,
@@ -76,6 +81,7 @@ import { GenericCameraManagerEngine } from '../generic/engine-generic';
 
 const EVENT_REQUEST_CACHE_MAX_AGE_SECONDS = 60;
 const RECORDING_SUMMARY_REQUEST_CACHE_MAX_AGE_SECONDS = 60;
+const MEDIA_METADATA_REQUEST_CACHE_AGE_SECONDS = 60;
 
 const CAMERA_BIRDSEYE = 'birdseye' as const;
 
@@ -802,14 +808,26 @@ export class FrigateCameraManagerEngine
   public async getMediaMetadata(
     hass: HomeAssistant,
     cameras: CameraConfigs,
-  ): Promise<MediaMetadata | null> {
+    query: MediaMetadataQuery,
+  ): Promise<MediaMetadataQueryResultsMap | null> {
+    const output: MediaMetadataQueryResultsMap = new Map();
+    if (this._requestCache.has(query)) {
+      const cachedResult = <MediaMetadataQueryResults | null>(
+        this._requestCache.get(query)
+      );
+      if (cachedResult) {
+        output.set(query, cachedResult as MediaMetadataQueryResults);
+        return output;
+      }
+    }
+
     const what: Set<string> = new Set();
     const where: Set<string> = new Set();
     const days: Set<string> = new Set();
 
     const instances = this._buildInstanceToCameraIDMapFromQuery(
       cameras,
-      new Set(cameras.keys()),
+      query.cameraIDs,
     );
 
     const processEventSummary = async (
@@ -865,14 +883,21 @@ export class FrigateCameraManagerEngine
       })(),
     );
 
-    if (!what.size && !where.size && !days.size) {
-      return null;
-    }
-    return {
-      ...(what.size && { what: what }),
-      ...(where.size && { where: where }),
-      ...(days.size && { days: days }),
+    const result: MediaMetadataQueryResults = {
+      type: QueryResultsType.MediaMetadata,
+      engine: Engine.Frigate,
+      metadata: {
+        ...(what.size && { what: what }),
+        ...(where.size && { where: where }),
+        ...(days.size && { days: days }),
+      },
+      expiry: add(new Date(), { seconds: MEDIA_METADATA_REQUEST_CACHE_AGE_SECONDS }),
+      cached: false,
     };
+
+    this._requestCache.set(query, { ...result, cached: true }, result.expiry);
+    output.set(query, result);
+    return output;
   }
 
   /**

--- a/src/camera-manager/frigate/types.ts
+++ b/src/camera-manager/frigate/types.ts
@@ -1,5 +1,11 @@
 import { z } from 'zod';
 import { dayToDate } from '../../utils/basic';
+import {
+  Engine,
+  EventQueryResults,
+  RecordingQueryResults,
+  RecordingSegmentsQueryResults,
+} from '../types';
 
 const dayStringToDate = (arg: unknown): Date | unknown => {
   return typeof arg === 'string' ? dayToDate(arg) : arg;
@@ -66,3 +72,25 @@ export const eventSummarySchema = z
   })
   .array();
 export type EventSummary = z.infer<typeof eventSummarySchema>;
+
+// ==============================
+// Frigate concrete query results
+// ==============================
+
+export interface FrigateEventQueryResults extends EventQueryResults {
+  engine: Engine.Frigate;
+  instanceID: string;
+  events: FrigateEvent[];
+}
+
+export interface FrigateRecordingQueryResults extends RecordingQueryResults {
+  engine: Engine.Frigate;
+  instanceID: string;
+  recordings: FrigateRecording[];
+}
+
+export interface FrigateRecordingSegmentsQueryResults
+  extends RecordingSegmentsQueryResults {
+  engine: Engine.Frigate;
+  instanceID: string;
+}

--- a/src/camera-manager/generic/engine-generic.ts
+++ b/src/camera-manager/generic/engine-generic.ts
@@ -23,6 +23,8 @@ import {
   CameraManagerCameraCapabilities,
   Engine,
   CameraEndpoints,
+  MediaMetadataQuery,
+  MediaMetadataQueryResultsMap,
 } from '../types';
 import { getEntityIcon, getEntityTitle } from '../../utils/ha';
 import { EntityRegistryManager } from '../../utils/ha/entity-registry';
@@ -139,7 +141,8 @@ export class GenericCameraManagerEngine implements CameraManagerEngine {
   public async getMediaMetadata(
     _hass: HomeAssistant,
     _cameras: CameraConfigs,
-  ): Promise<MediaMetadata | null> {
+    _query: MediaMetadataQuery,
+  ): Promise<MediaMetadataQueryResultsMap | null> {
     return null;
   }
 

--- a/src/camera-manager/manager.ts
+++ b/src/camera-manager/manager.ts
@@ -30,6 +30,7 @@ import {
   RecordingSegmentsQueryResultsMap,
   ResultsMap,
   CameraEndpoints,
+  Engine,
 } from './types.js';
 import orderBy from 'lodash-es/orderBy';
 import { CameraManagerEngineFactory } from './engine-factory.js';
@@ -106,26 +107,39 @@ export class CameraManager {
     this._store = new CameraManagerStore();
   }
 
+  protected async _getEnginesForCameras(
+    hass: HomeAssistant,
+    camerasConfig: CamerasConfig,
+  ): Promise<Map<CameraConfig, CameraManagerEngine>> {
+    const output: Map<CameraConfig, CameraManagerEngine> = new Map();
+    const engines: Map<Engine, CameraManagerEngine> = new Map();
+
+    for (const cameraConfig of camerasConfig) {
+      const engineType = await this._engineFactory.getEngineForCamera(
+        hass,
+        cameraConfig,
+      );
+      const engine = engineType
+        ? engines.get(engineType) ?? this._engineFactory.createEngine(engineType)
+        : null;
+      if (!engine || !engineType) {
+        throw new CameraInitializationError(
+          localize('error.no_camera_engine'),
+          cameraConfig,
+        );
+      }
+      engines.set(engineType, engine);
+      output.set(cameraConfig, engine);
+    }
+    return output;
+  }
+
   protected async _initializeCamera(
     hass: HomeAssistant,
+    engine: CameraManagerEngine,
     entityRegistryManager: EntityRegistryManager,
     inputCameraConfig: CameraConfig,
   ): Promise<InitializedCamera> {
-    const engineType = await this._engineFactory.getEngineForCamera(
-      hass,
-      inputCameraConfig,
-    );
-    const engine = engineType
-      ? this._store.getEngineOfType(engineType) ??
-        (await this._engineFactory.createEngine(engineType))
-      : null;
-    if (!engine) {
-      throw new CameraInitializationError(
-        localize('error.no_camera_engine'),
-        inputCameraConfig,
-      );
-    }
-
     const initializedConfig = await engine.initializeCamera(
       hass,
       entityRegistryManager,
@@ -159,10 +173,15 @@ export class CameraManager {
       await entityRegistryManager.fetchEntityList(hass);
     }
 
+    // Engines are created sequentially, to avoid duplicate creation of the same
+    // engine. See: https://github.com/dermotduffy/frigate-hass-card/issues/941
+    const engineByConfig = await this._getEnginesForCameras(hass, camerasConfig);
+
+    // Configuration is initialized in parallel.
     const results = await allPromises(
-      camerasConfig,
-      async (cameraConfig) =>
-        await this._initializeCamera(hass, entityRegistryManager, cameraConfig),
+      engineByConfig.entries(),
+      async ([cameraConfig, engine]) =>
+        await this._initializeCamera(hass, engine, entityRegistryManager, cameraConfig),
     );
 
     // Do the additions based off the result-order, to ensure the map order is

--- a/src/camera-manager/manager.ts
+++ b/src/camera-manager/manager.ts
@@ -31,6 +31,8 @@ import {
   ResultsMap,
   CameraEndpoints,
   Engine,
+  MediaMetadataQuery,
+  MediaMetadataQueryResults,
 } from './types.js';
 import orderBy from 'lodash-es/orderBy';
 import { CameraManagerEngineFactory } from './engine-factory.js';
@@ -62,6 +64,11 @@ class QueryClassifier {
   ): query is RecordingSegmentsQuery {
     return query.type === QueryType.RecordingSegments;
   }
+  public static isMediaMetadataQuery(
+    query: DataQuery | PartialDataQuery,
+  ): query is MediaMetadataQuery {
+    return query.type === QueryType.MediaMetadata;
+  }
 }
 
 class QueryResultClassifier {
@@ -79,6 +86,11 @@ class QueryResultClassifier {
     queryResults: QueryResults,
   ): queryResults is RecordingSegmentsQueryResults {
     return queryResults.type === QueryResultsType.RecordingSegments;
+  }
+  public static isMediaMetadataQuery(
+    queryResults: QueryResults,
+  ): queryResults is MediaMetadataQueryResults {
+    return queryResults.type === QueryResultsType.MediaMetadata;
   }
 }
 
@@ -268,27 +280,24 @@ export class CameraManager {
     const where: Set<string> = new Set();
     const days: Set<string> = new Set();
 
-    const engines = this._store.getAllEngines();
-
-    const processMetadata = async (engine: CameraManagerEngine): Promise<void> => {
-      const engineMetadata = await engine.getMediaMetadata(
-        hass,
-        this._store.getCameras(),
-      );
-      if (engineMetadata) {
-        if (engineMetadata.what) {
-          engineMetadata.what.forEach(what.add, what);
-        }
-        if (engineMetadata.where) {
-          engineMetadata.where.forEach(where.add, where);
-        }
-        if (engineMetadata.days) {
-          engineMetadata.days.forEach(days.add, days);
-        }
-      }
+    const query: MediaMetadataQuery = {
+      type: QueryType.MediaMetadata,
+      cameraIDs: this._store.getCameraIDs(),
     };
 
-    await allPromises(engines, processMetadata);
+    const results = await this._handleQuery(hass, query);
+
+    for (const [query, result] of results?.entries() ?? []) {
+      if (result.metadata.what) {
+        result.metadata.what.forEach(what.add, what);
+      }
+      if (result.metadata.where) {
+        result.metadata.where.forEach(where.add, where);
+      }
+      if (result.metadata.days) {
+        result.metadata.days.forEach(days.add, days);
+      }
+    }
 
     if (!what.size && !where.size && !days.size) {
       return null;
@@ -565,6 +574,12 @@ export class CameraManager {
         )) as Map<QT, QueryReturnType<QT>> | null;
       } else if (QueryClassifier.isRecordingSegmentsQuery(query)) {
         engineResult = (await engine.getRecordingSegments(
+          hass,
+          this._store.getCameras(),
+          query,
+        )) as Map<QT, QueryReturnType<QT>> | null;
+      } else if (QueryClassifier.isMediaMetadataQuery(query)) {
+        engineResult = (await engine.getMediaMetadata(
           hass,
           this._store.getCameras(),
           query,

--- a/src/camera-manager/store.ts
+++ b/src/camera-manager/store.ts
@@ -1,4 +1,3 @@
-import uniq from 'lodash-es/uniq';
 import { CameraConfig } from '../types';
 import { ViewMedia } from '../view/media';
 import { CameraManagerEngine } from './engine';
@@ -8,7 +7,7 @@ type CameraManagerEngineCameraIDMap = Map<CameraManagerEngine, Set<string>>;
 
 export class CameraManagerStore {
   protected _configs: Map<string, CameraConfig> = new Map();
-  protected _engines: Map<string, CameraManagerEngine> = new Map();
+  protected _enginesByCamera: Map<string, CameraManagerEngine> = new Map();
   protected _enginesByType: Map<Engine, CameraManagerEngine> = new Map();
 
   public addCamera(
@@ -17,7 +16,7 @@ export class CameraManagerStore {
     engine: CameraManagerEngine,
   ): void {
     this._configs.set(cameraID, cameraConfig);
-    this._engines.set(cameraID, engine);
+    this._enginesByCamera.set(cameraID, engine);
     this._enginesByType.set(engine.getEngineType(), engine);
   }
 
@@ -54,7 +53,7 @@ export class CameraManagerStore {
   }
 
   public getEngineForCameraID(cameraID: string): CameraManagerEngine | null {
-    return this._engines.get(cameraID) ?? null;
+    return this._enginesByCamera.get(cameraID) ?? null;
   }
 
   public getEnginesForCameraIDs(
@@ -84,6 +83,6 @@ export class CameraManagerStore {
   }
 
   public getAllEngines(): CameraManagerEngine[] {
-    return uniq([...this._engines.values()]);
+    return [...this._enginesByType.values()];
   }
 }

--- a/src/camera-manager/types.ts
+++ b/src/camera-manager/types.ts
@@ -1,6 +1,5 @@
 import { CameraConfig, FrigateCardView } from '../types';
 import { ViewMedia } from '../view/media';
-import { FrigateEvent, FrigateRecording } from './frigate/types';
 
 // ====
 // Base
@@ -10,12 +9,14 @@ export enum QueryType {
   Event = 'event-query',
   Recording = 'recording-query',
   RecordingSegments = 'recording-segments-query',
+  MediaMetadata = 'media-metadata',
 }
 
 export enum QueryResultsType {
   Event = 'event-results',
   Recording = 'recording-results',
   RecordingSegments = 'recording-segments-results',
+  MediaMetadata = 'media-metadata-results',
 }
 
 export enum Engine {
@@ -65,6 +66,8 @@ export type QueryReturnType<QT> = QT extends EventQuery
   ? RecordingQueryResults
   : QT extends RecordingSegmentsQuery
   ? RecordingSegmentsQueryResults
+  : QT extends MediaMetadataQuery
+  ? MediaMetadataQueryResults
   : never;
 export type PartialQueryConcreteType<PQT> = PQT extends PartialEventQuery
   ? EventQuery
@@ -78,6 +81,7 @@ export type ResultsMap<QT> = Map<QT, QueryReturnType<QT>>;
 export type EventQueryResultsMap = ResultsMap<EventQuery>;
 export type RecordingQueryResultsMap = ResultsMap<RecordingQuery>;
 export type RecordingSegmentsQueryResultsMap = ResultsMap<RecordingSegmentsQuery>;
+export type MediaMetadataQueryResultsMap = ResultsMap<MediaMetadataQuery>;
 
 export interface MediaMetadata {
   where?: Set<string>;
@@ -176,24 +180,15 @@ export interface RecordingSegmentsQueryResults extends QueryResults {
   segments: RecordingSegment[];
 }
 
-// ========================
-// Frigate concrete results
-// ========================
+// ====================
+// Media metadata Query
+// ====================
 
-export interface FrigateEventQueryResults extends EventQueryResults {
-  engine: Engine.Frigate;
-  instanceID: string;
-  events: FrigateEvent[];
+export interface MediaMetadataQuery extends DataQuery {
+  type: QueryType.MediaMetadata;
 }
 
-export interface FrigateRecordingQueryResults extends RecordingQueryResults {
-  engine: Engine.Frigate;
-  instanceID: string;
-  recordings: FrigateRecording[];
-}
-
-export interface FrigateRecordingSegmentsQueryResults
-  extends RecordingSegmentsQueryResults {
-  engine: Engine.Frigate;
-  instanceID: string;
+export interface MediaMetadataQueryResults extends QueryResults {
+  type: QueryResultsType.MediaMetadata;
+  metadata: MediaMetadata;
 }

